### PR TITLE
propagate image env/dir to `exec`

### LINKED ISSUE #3066: Use image config to configure action ExecOps
Right now any env vars (like `PATH`) or other configuration provided as part of an image config w/ the `FS` for a package don't get set when invoking the action, resulting in hacks like this: https://github.com/dagger/cloak/blob/e60aede62917e4e2711cca499d019c857f0b14ed/examples/netlify/ts/index.ts#L17-L17

### PR DESCRIPTION
**Pull request by vito**
fixes #3066

supersedes https://github.com/dagger/cloak/pull/220 which had a totally different approach

Introduces a new intermediate type, `filesystem.Image`, which encodes to/from `FSID`. (Name bikeshedding welcome.) This type pairs the LLB definition with an OCI image config (`{"Env":[...],"Dir":"...",...}`). It can be converted to and from an `llb.State` via the `Env`/`Dir` fields stored within. `llb.Image` is now passed `llb.WithMetaResolver` in order to populate these fields.

I'm not sure if this is considered backwards-incompatible; within a single engine session nothing should break, but the structure of `FSID` has changed, so any previously-generated `FSID`s won't work with this new code (and vice versa),

TODO:

* [x] test `Env` propagation from image
* [x] test `Dir` propagation from image

### CODE REVIEW

File: core/core.go [hunk 1]
```diff
@@ -62,20 +62,20 @@ type baseSchema struct {
...
 	_, err = r.gw.Solve(ctx, bkgw.SolveRequest{
 		Evaluate:   true,
-		Definition: def.ToPB(),
+		Definition: info.FS,
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	// FIXME: should we create a filesystem from `res.SingleRef()`?
```
Comment (vito) [inline core/core.go:77]: Rambling/thinking/learning out loud, posting for posterity:

**tl;dr: I'll refactor the code a bit to respect this `FIXME` since we're using `SingleRef` everywhere else and it should also shrink `FSID` a bit.**

I originally removed this `FIXME` because when I tried going from `res.SingleRef` -> `ref.ToState` -> `filesystem.FromState` the image config would be lost from the original `st`.

But in hindsight that's an orthogonal problem not made any more or less possible with my change. We're just using `llb.State` as a kind of brittle vessel for this config, so I'd just need to pluck the env/dir from the input `llb.State` instead of the returned `llb.State`.

I'm not sure what `SingleRef` is actually for so I'm doing a bit of investigation first. From what I gather it seems to be a way to refer to a previously built value instead of its source LLB definition. Is this an optimization or does it have any observable behavior?

I did a gross hack to get this to use the config from the input `llb.State` and the filesystem from the output `llb.State` (it'll be less gross after I refactor) and all the tests pass. I also made it log when the marshalled value is different, and indeed it's a few kilobytes smaller sometimes, but it's also pretty frequently the same.

```patch
diff --git a/core/core.go b/core/core.go
index 758e375a..bef7402c 100644
--- a/core/core.go
+++ b/core/core.go
@@ -1,7 +1,9 @@
 package core
 
 import (
+	"bytes"
 	"context"
+	"log"
 
 	bkclient "github.com/moby/buildkit/client"
 	"github.com/moby/buildkit/client/llb"
@@ -62,20 +64,62 @@ type baseSchema struct {
 }
 
 func (r *baseSchema) Solve(ctx context.Context, st llb.State, marshalOpts ...llb.ConstraintsOpt) (*filesystem.Filesystem, error) {
-	info, err := filesystem.ImageFromState(ctx, st, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	env, err := st.Env(ctx)
 	if err != nil {
 		return nil, err
 	}
+	dir, err := st.GetDir(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	input, err := st.Marshal(ctx, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	if err != nil {
+		return nil, err
+	}
+
+	ibytes, _ := input.ToPB().Marshal()
 
-	_, err = r.gw.Solve(ctx, bkgw.SolveRequest{
+	res, err := r.gw.Solve(ctx, bkgw.SolveRequest{
 		Evaluate:   true,
-		Definition: info.FS,
+		Definition: input.ToPB(),
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	return info.ToFilesystem()
+	bkref, err := res.SingleRef()
+	if err != nil {
+		return nil, err
+	}
+
+	ost, err := bkref.ToState()
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := ost.Marshal(ctx, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	if err != nil {
+		return nil, err
+	}
+
+	obytes, _ := output.ToPB().Marshal()
+
+	if bytes.Equal(ibytes, obytes) {
+		log.Println("!!! INPUT/OUTPUT SAME", len(ibytes), len(obytes))
+	} else {
+		dir, _ := os.MkdirTemp("/tmp/results", "result*")
+		os.WriteFile(filepath.Join(dir, "input"), ibytes, 0644)
+		os.WriteFile(filepath.Join(dir, "output"), obytes, 0644)
+		fmt.Println(string(ibytes), string(obytes))
+		log.Println("!!! INPUT/OUTPUT DIFF", len(ibytes), len(obytes))
+	}
+
+	return (&filesystem.Image{
+		FS: output.ToPB(),
+		Config: specs.ImageConfig{
+			Env:        env,
+			WorkingDir: dir,
+		},
+	}).ToFilesystem()
+	// return filesystem.FromState(ctx, st, r.platform)
+	// return info.ToFilesystem()
 }
 
 func (r *baseSchema) Export(ctx context.Context, fs *filesystem.Filesystem, export bkclient.ExportEntry) error {
```

Some samples from `go test -v ./...`:

```
2022/09/26 13:07:02 !!! INPUT/OUTPUT DIFF 9012 6674
2022/09/26 13:07:02 !!! INPUT/OUTPUT SAME 7269 7269
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 7278 7278
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 914 914
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 550 550
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 914 914
2022/09/26 13:07:04 !!! INPUT/OUTPUT SAME 550 550
2022/09/26 13:07:05 !!! INPUT/OUTPUT DIFF 13311 9966
2022/09/26 13:07:05 !!! INPUT/OUTPUT SAME 914 914
2022/09/26 13:07:05 !!! INPUT/OUTPUT SAME 557 557
```

The difference in the marshaled protobuf is always that the shim `Dockerfile` and `main.go` are removed:

image

I'm not 100% sure what the rules are for what gets omitted in the returned result, but yeah it's a bit smaller.

I'll refactor the code a bit to respect this comment since we're using it everywhere else. It should also shrink `FSID` a bit.

File: core/core.go [hunk 2]
```diff
@@ -62,20 +62,42 @@ type baseSchema struct {
...
+	if err != nil {
+		return nil, err
+	}
+
+	resSt, err := bkref.ToState()
+	if err != nil {
+		return nil, err
+	}
+
+	img, err := filesystem.ImageFromStateAndConfig(ctx, resSt, inputCfg, marshalOpts...)
```
Comment (vito) [inline core/core.go:95]: I'm not sure if this actually needs `marshalOpts` since we already used it above and now we're just marshaling the result. :thinking:

**Comment by vito** [comment]
@sipsma This is ready for another look whenever you have time!

Latest changes fix up tests and address a `FIXME` that I had previously just deleted. There's a giant resolved comment above if you're curious about the thought behind it, but feel free to skip.

**Review by sipsma — APPROVED** [review]
LGTM, the only things that stuck out to me are things that we will be improving in the new API refactor anyways, so this looks great as is, thank you!
